text/0800-ts-adoption-plan.md

```diff
@@ -0,0 +1,833 @@
+    import Component from '@ember/component';
+    import { service } from '@ember/service';
+    import { action } from '@ember/object';
+
+    export default Component.extend({
+      // session here has the type SessionService...
+      session: service('session'),
+    }) {
+      @action login(username: string, password: string) {
+        // ...so this call is type safe
+        this.session.login(username, string);
+      }
+    }
+    ```
+
+With the introduction of full support for native classes and decorators in Octane, there is no need for this (and it was rather janky to begin with!). Accordingly, we will simply remove support for these two type registries entirely:
+
+- The service `Registry` in `@ember/service`
+- The controller `Registry` in `@ember/controller`
+
+The decorators will continue to accept a string key, but will not validate that against a registry, since decorators cannot change the types of the items they decorator in TypeScript.[^reintroduce]
+
+For Ember Data, there is considerably more ongoing need for registry-style APIs. `Store.findRecord` isn’t going anywhere any time soon, for example, and it *requires* some kind of registry to make `this.store.findRecord('user', 1)` correctly return a `User` model. Many other APIs within Ember Data similarly require registries for type safety. Thus, Ember Data will need its own dedicated design for handling those, and this RFC leaves those for the Ember Data team to address in a dedicated RFC.
+
+[^reintroduce]: We could conceivably reintroduce a registry of this form in the future if that limitation on decorators in TypeScript changes, so that `@service('session') declare session;` would have the right type without needing an explicit type annotation, but that would need to be well-motivated on the merits, and there are actually advantages to having explicit type imports for the static analyzability of the code. The combination of Embroider and other future design moves we might make for the design system, for example [this experiment by pzuraq](https://github.com/pzuraq/ember-totally-not-module-based-services), may make other directions more appealing than the registry approach.
+
+
+### CLI Integration
+
+We will introduce a new `--typescript` (`-ts`) flag for the `ember new` and `ember addon` commands, allowing users to opt into TypeScript when generating a new project. Using this flag will:
+
+- Install Ember TypeScript integration tooling, today consisting simply of `ember-cli-typescript` and its generators.
+- Set the `isTypeScriptProject` option for `.ember-cli` (introduced in [RFC #0776][rfc-0776]) to `true`, so that blueprints are generated with TypeScript by default in the project.
+- Configure linting:
+    - In the`.eslintrc.js` blueprint:
+        - use `@typescript-eslint/parser` instead of the Babel ESLint parser
+        - include `@typescript-eslint` in the `plugins` key
+        - include `plugin:@typescript-eslint/recommended` in the `extends` key
```

Comment:
To add some more context:
* https://typescript-eslint.io/docs/linting/type-linting/
* https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/recommended-requiring-type-checking.ts

Reply:
@simonihmig the Ember TypeScript team has *not* generally recommended using those. While it’s nice that they exist, they are (as you say) *dramatically* slower than regular ESLint rules, because they have to invoke the compiler. That’s a pretty big experiential difference for people, and it hasn’t (to date) been clear that it’s worth doing that out of the box, so we have avoided recommending them by default. (Also because it's quite easy to add them later if you want them!)

Reply:
To clarify, I distinguish strongly between *type checking*, which should be as strict as possible, and *linting*, which has a lot more flexibility in terms of what makes sense as a baseline. Different teams and projects need different things there. I’ll also clarify: I really mean “it’s easy to add them at setup time yourself if you want them”—certainly adding lint rules later can be non-trivial amounts of work!

One last note: the default config from ember-cli-typescript has, as far as I remember, always included `noImplicitAny`: we’ve always tried to enable every possible strict flag by default, the only historical exception being `strictFunctionTypes`, which our attempts to work with mixin types and `get`/`set` could not handle. It’s quite possible you’re thinking of the ESLint `no-explicit-any` rule, though. I strongly recommend that one! Also it’s in the baseline and doesn’t require type checking. 😉